src/registry2.0/portimage.tcl

```diff
@@ -349,9 +355,17 @@ proc extract_archive_to_tmpdir {location} {
                 set tar "tar"
                 if {[catch {set tar [macports::findBinary $tar ${macports::autoconf::tar_path}]} errmsg] == 0} {
                     # Opportunistic HFS compression. bsdtar will automatically
-                    # ignore if on non-HFS filesystem.
+                    # ignore if on non-HFS filesystem, or if extraction device
+                    # is not the same as prefix / applications_dir
+                    # The comparisons against prefix and applications_dir should be
+                    # revisted if rsync or ditto is eventually used to perform final
+                    # move into destinations.
+                    set extrfs [_getfs ${extractdir}]
                     if {![catch {macports::binaryInPath bsdtar}] &&
-                        ![catch {exec bsdtar -x --hfsCompression < /dev/null >& /dev/null}]} {
+                        ![catch {exec bsdtar -x --hfsCompression < /dev/null >& /dev/null}] &&
+                        ${extrfs} == [_getfs ${macports::prefix}] &&
+                        ${extrfs} == [_getfs ${macports::applications_dir}] &&
+                        [string length [exec mount | egrep "${extrfs} .*(hfs|apfs)"]] > 0 } {
```

Comment:
The match needs to be anchored at the beginning of the line with `^` and needs to match the correct field. What if the mount point directory contains the substring "hfs" or "apfs"?

Also see above, should rather use `statfs(2)` to determine filesystem.

Reply:
Yep. Will look at this weekend as time permits.